src/dataflow/source/kafka.rs

```diff
@@ -68,42 +68,108 @@ impl<'a> From<&BorrowedMessage<'a>> for MessageParts {
-                    false
-                } else {
-                    true
+    let mut result = true;
+
+    for c in consumers {
+        match c.fetch_metadata(Some(topic), Duration::from_secs(1)) {
+            Ok(md) => match md.topics().iter().find(|mdt| mdt.name() == topic) {
+                None => {
+                    warn!("Topic {} not found in Kafka metadata", topic);
+                    result = false;
+                    break;
                 }
+                Some(mdt) => {
+                    let partitions = mdt.partitions().len();
+                    if partitions < expected_partitions {
+                        warn!(
+                            "Topic {} does not have as many partitions as expected ({} < {})",
+                            topic, partitions, expected_partitions
+                        );
+                        result = false;
+                        break;
+                    }
+                }
+            },
+            Err(e) => {
+                warn!("Error refreshing Kafka metadata: {}", e);
+                result = false;
+                break;
             }
-        },
-        Err(e) => {
-            warn!("Error refreshing Kafka metadata: {}", e);
-            false
         }
-    };
+    }
+
     if !result {
         warn!("Descheduling for 1s to wait for Kafka metadata to catch up");
         activator.activate_after(Duration::from_secs(1));
     }
     result
 }
 
+/// Create a new kafka consumer
+fn create_kafka_consumer(
+    activator: Arc<Mutex<SyncActivator>>,
+    topic: &str,
+    kafka_config: &ClientConfig,
+) -> BaseConsumer<GlueConsumerContext> {
+    let cx = GlueConsumerContext(activator);
+    let consumer: BaseConsumer<GlueConsumerContext> = kafka_config
+        .create_with_context(cx)
+        .expect("Failed to create Kafka Consumer");
+    consumer.subscribe(&[topic]).unwrap();
+    consumer
+}
+
+/// Update the list of Kafka consumers to match the number of partitions
+/// We currently create one customer per partition
+fn update_consumer_list(
+    consumers: &mut VecDeque<BaseConsumer<GlueConsumerContext>>,
+    topic: &str,
+    to_add: usize,
+    activator: Arc<Mutex<SyncActivator>>,
+    kafka_config: &ClientConfig,
+) {
+    for _ in 0..to_add {
+        consumers.push_front(create_kafka_consumer(
+            activator.clone(),
+            topic,
+            kafka_config,
+        ));
+    }
+}
+
+/// Poll from the next customer for which a message is available. This function polls the set
+/// round-robin: when a consumer is polled, it is placed at the back of the queue.
+fn get_next_message_from_consumers(
+    name: &str,
+    consumers: &mut VecDeque<BaseConsumer<GlueConsumerContext>>,
+) -> Option<(MessageParts, BaseConsumer<GlueConsumerContext>)> {
+    let consumer_count = consumers.len();
+    let mut attempts = 0;
+    // No currently buffered message, poll from stream
+    while attempts < consumer_count {
```

Comment:
Not a blocker, but if this was just a vec of consumers could you just `for consumer in consumers.iter_mut()`, or does something here need ownership?

Reply:
It may return half-way through, and wants to leave the queue in a state where on the next invocation it starts where it left off. Doing it from zero each time would return to the non-round-robin implementation, I think.

Reply:
ah yeah completely fair

Reply:
Is this a pun? :D